File: Java/src/Stack/ArrayBaseStack.java
package Stack;

import DynamicArray.DynamicArray;

import java.util.EmptyStackException;
import java.util.Iterator;

public class ArrayBaseStack<T> implements StackADT<T> {
    private final DynamicArray<T> array;
    private int index = -1;

    public ArrayBaseStack(int initSize) {
        array = new DynamicArray<T>(initSize);
    }


    @Override
    public void push(T element) {
        index++;
        array.add(element);
    }

    @Override
    public T pop() {
        if (isEmpty()) {
            throw new EmptyStackException();
        } else {
            return array.removeAtWithoutMoving(index--);
        }
    }

    @Override
    public T top() {
        return array.get(index);
    }

    @Override
    public int size() {
        return array.size();
    }

    @Override
    public boolean isEmpty() {
        return array.isEmpty();
    }

    @Override
    public Iterator<T> iterator() {
        return array.iterator();
    }
}

File: Typescript/Stacks/ArrayBaseStack.ts
import { DynamicArray } from "../DynamicArray/DynamicArray";
import { StackADT } from "./StackADT";

export class ArrayBaseStack<T> implements StackADT<T> {
  private array: DynamicArray<T>;
  private index: number = -1;

  constructor(initSize: number) {
    this.array = new DynamicArray<T>(initSize);
  }

  public push(value: T): void {
    this.index++;
    this.array.add(value);
  }

  public pop(): T | null {
    if (this.array.isEmpty()) {
      throw new Error("Empty Stack!");
    }
    return this.array.removeAtWithoutMoving(this.index--);
    // The stack only interacts with the first element because of the FIFO mechanism
    // The index -1 because of array starting from the 0
  }

  public top(): T | null {
    return this.array.getAt(this.index);
  }

  public size(): number {
    return this.array.length();
  }

  public isEmpty(): boolean {
    return this.array.isEmpty();
  }
}
